include/kernel.h

```diff
@@ -3959,6 +3976,219 @@ static inline char *K_THREAD_STACK_BUFFER(k_thread_stack_t sym)
 
 #endif /* _ARCH_DECLARE_STACK */
 
+/**
+ * @defgroup mem_domain_apis Memory domain APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/** @def MEM_PARTITION_ENTRY
+ *  @brief Used to declare a memory partition entry
+ */
+#define MEM_PARTITION_ENTRY(_start, _size, _attr) \
+	{\
+		.start = _start, \
+		.size = _size, \
+		.attr = _attr, \
+	}
+
+/** @def K_MEM_PARTITION_DEFINE
+ *  @brief Used to declare a memory partition
+ */
+#ifdef _ARCH_MEM_PARTITION_ALIGN_CHECK
+#define K_MEM_PARTITION_DEFINE(name, start, size, attr) \
+	_ARCH_MEM_PARTITION_ALIGN_CHECK(start, size); \
+	struct k_mem_partition name = \
+		MEM_PARTITION_ENTRY((u32_t)start, size, attr)
+#else
+#define K_MEM_PARTITION_DEFINE(name, start, size, attr) \
+	struct k_mem_partition name = \
+		MEM_PARTITION_ENTRY((u32_t)start, size, attr)
+#endif /* _ARCH_MEM_PARTITION_ALIGN_CHECK */
+
+/* memory partition arch/soc independent attribute */
+/** @def K_MEM_PARTITION_P_NA_U_NA
+ *  @brief Denote partition is "Privileged No Access, Unprivileged No Access"
+ */
+#define K_MEM_PARTITION_P_NA_U_NA				0
```

Comment:
I think K_MEM_PARTITION_P_NA_U_NA could be done by setting page table entry to 'not present'.
The K_MEM_PARTITION_P_RW_U_RO could be done if the WP bit of CR0 is cleared (allow supervisor to write a read-only page) and the R bit of the page table entry is cleared too (a read-only page). But this would cause K_MEM_PARTITION_P_RO_U_RO can't be specified since supervisor could access all read-only pages if WP bit of CR0 is cleared. Please correct me if I'm wrong.

Since not all attributes could be implemented on x86, I think I should follow your suggestion to move these attributes to arch.h.